ARROW-2704: [Java] Change MessageReader API to improve custom message handling for streams

BryanCutler opened the pull request:
Made a number of improvements to make stream processing for messages easier without having to load Arrow record and dictionary batches

### Changes Made

- Changed `MessageReader` interface to have a `readNext` method that takes an extension of `MessageHolder` to return the message data. See details below.

- Added static function in ArrowStreamWriter to write the EOS identifier

- Moved intToBytes to MessageSerializer (where bytesToInt is), and now works with an existing byte array

- Removed the abstract identifier in beginInternal and endInternal of `ArrowReader` so that subclasses are not forced to implement if they are not needed.

- Made deserialze* functions in `MessageSerializer` more consistent.  In general, each can deserialze from a `ReadChannel` or from data directly.

### MessageReader API

The changes to `MessageReader` are to make it easy to subclass and implement custom handling of reading messages/data.  `readNext` input is a `MessageHolder` instance which can be extended to store message data and additional information during the read.

### Usage in ArrowStreamReader

There is one implementation of `MessageReader`, `MessageChannelReader` which reads the body data into an offheap `ArrowBuf`. `ArrowStreamReader` uses this implementation to do the message reading and then handles the deserialization of batch data into vectors.

### Testing

Updated existing tests for new APIs, added test for write/read of zero-length record batch and for message alignment.

BryanCutler commented:
These changes were pretty useful when working on https://github.com/apache/spark/pull/20725
@icexelloss @siddharthteotia please take a look if you can, thanks!

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelReader.java:
```diff
@@ -61,13 +61,18 @@ public Message readNextMessage() throws IOException {
+  /**
+   * Read a Message of the given length into the existing buffer and return the loaded Message.
+   */
+  protected Message loadMessage(int messageLength, ByteBuffer buffer) throws IOException {
```
Bryan, can you give a bit context why do you want to provide your own buffer when reading the message?

BryanCutler replied:
So `loadMessage` here will create an Arrow `Message` of the given `messageLength` from the `buffer`.  What I would like to be able to do is override this method to then have access to the message buffer.  This way I can examine the message contents and then copy it to another buffer without having to serialize it again.  For example (in Scala):

```scala
override protected def loadMessage(messageLength: Int, buffer: ByteBuffer): Message = {
        val msg = super.loadMessage(messageLength, buffer)
        val bodyLength = msg.bodyLength().asInstanceOf[Int]

        if (msg.headerType() == MessageHeader.RecordBatch) {
          val allbuf = ByteBuffer.allocate(4 + messageLength + bodyLength)
          allbuf.put(WriteChannel.intToBytes(messageLength))
          allbuf.put(buffer)
          in.readFully(allbuf)
          lastBatch = allbuf.array()
        } else {...}

        msg
```

this allows me to create a new buffer with the existing message and then read the record batch body directly into the new buffer.  So I end up with a single buffer containing the message + body with minimal processing/copying.

icexelloss replied:
I see, you want to serialize an arrow record batch message (metadata + body) into a Java on-heap memory rather than Arrow off-heap memory?

icexelloss replied:
I think your loadMessage message also changes the semantics of the function from "reading metadata" to "reading metadata and body" in the subclass...

I am not sure it's a good idea from an OOP perspective..

BryanCutler replied:
>I see, you want to serialize an arrow record batch message (metadata + body) into a Java on-heap memory rather than Arrow off-heap memory?

Sort of.. I need to read a Arrow record batch message from an input stream to a byte array.  I do not need to look at the data and must pass on ownership of the memory,  so I want to avoid loading to Arrow buffers or making unneeded copies.

> I think your loadMessage message also changes the semantics of the function from "reading metadata" to "reading metadata and body" in the subclass...
I am not sure it's a good idea from an OOP perspective..

Nothing changes in `MessageChannelReader`, and the details of a subclass to it aren't really relevant here.  From an Arrow perspective, these classes should provide utility to process streams in a flexible way.  I'm up for any changes if you have a suggestion that is just as efficient and can continue to work with ArrowStreamReader.

icexelloss replied:
@BryanCutler how about we discuss on https://github.com/apache/spark/pull/21546/files/a5a1fbe7121c5b0dd93876a56c29ad17dcd9b168#r194898793 first and then decide what to do here?

BryanCutler commented:
**Update**

I made some changes to `MessageReader` that will solve the issues I've been dealing with.  I'll do a better write up later, but here is the gist of the latest changes:  `MessageReader` is now an abstract class with a single public method `readNext`. This can be called repeatably and return false when no more messages in the stream. It takes in a `MessageReadHolder` class that will have variables populated with all data used in the reading of a message, like length, the raw buffer and `Message`.  There are 2 abstract methods `readMessage` and `readBody` that are invoked by `readNext`.

There is one concrete implementation of `MessageReader` that is `MessageChannelReader` that implements the above methods with input from a `ReadChannel` and loads message body data into an `ArrowBuf`.  This reader is used by `ArrowStreamReader` to read messages/data and deserialize into Schema and Record/DictionaryBatches.

When making a custom `MessageReader` you can extend the `MessageReadHolder` as needed and implement the `readMessage` and `readMessageBody` methods to populate it.

BryanCutler commented:
@icexelloss I'll do a more thorough writeup and probably add some tests later, but if you could take a look and let me know your thoughts on this approach, I'd appreciate it.  Thanks!

Also, since there are some API changes, maybe best to send out to the dev list to get feedback..

icexelloss commented:
@BryanCutler Could you please sum up the purpose of the API change?

BryanCutler commented:
Thanks @elahrvivaz !  @icexelloss I'll add a summary of the API change in the description

icexelloss commented:
@BryanCutler Thanks! There seems to be fair amount of API changes in the MessageSerializer class. It would be good to understand better the rational behind the API changes.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -56,19 +56,76 @@
+   * @return
+   * @throws IOException
+   */
+  public static int writeMessageBufferAligned(WriteChannel out, int messageLength, ByteBuffer messageBuffer) throws IOException {
```
TODO: need tests

BryanCutler replied:
done

wesm replied:
Would you ever write an unaligned message (it is not permitted by the message protocol)? I question whether including details of the protocol in the function name is a good idea

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelReader.java:
```diff
@@ -18,98 +18,70 @@
-  public MessageChannelReader(ReadChannel in) {
+  public MessageChannelReader(ReadChannel in, BufferAllocator allocator) {
     this.in = in;
+    this.allocator = allocator;
```
This somewhat change the ownership of the arrow memory allocated here from the caller to the reader instance. I feel this maybe an API that needs some more discussion.

Specifically, before the change, when a caller uses this class to read arrow data, the memory ownership is pretty clear - the caller owners the BufferAllocator. However, now the allocator is inside the class and therefore not owned by the caller, does it mean the ownership of the data is now changed? Or does the caller still own the memory, but not the allocator? I found this a little bit hard to reason...

cc @jacques-n WDYT?

BryanCutler replied:
I don't think this really changes memory ownership.  The `MessageChannelReader` is constructed with a given allocator, it does not create one.  The allocated buffers are owned by the creator of the `MessageChannelReader` instance, which seems like how it should be. And usually the creator of the instance is going to be the same that calls `readNext()`, where the allocation is performed.  I think it's unlikely that one would change allocators from message to message, so this makes things easier.

icexelloss replied:
I see. This seems fine.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReadHolder.java:
```diff
@@ -0,0 +1,34 @@
+ * The variables are defined as follows: messageLength is the length of the message read,
+ * messageBuffer contains the raw bytes of the message, message is the realized flatbuf Message.
+ */
+public class MessageReadHolder {
```
Should we have an interface for this?

BryanCutler replied:
I don't think so, this represents the minimum information gathered when building a FB message.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReader.java:
```diff
@@ -18,48 +18,50 @@
+ * Abstract base class for reading a sequence of messages.
  */
-public interface MessageReader {
+public abstract class MessageReader<T extends MessageReadHolder> {
```
Should we keep this an interface?

BryanCutler replied:
The point of making this an abstract class was to include the logic of the Arrow stream format that a message is prefixed by a length, then has an optional body determined by a field in the message.

BryanCutler commented:
Thanks for the review @icexelloss, I'll ping @siddharthteotia again to see if he has any input on this.

BryanCutler commented:
I'll leave this open for another day and merge it on Friday if no objections

wesm commented:
Thanks @BryanCutler. Since we're so close to 0.10.0 if there are any changes required obviously it would be good to address them now

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReadHolder.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public class MessageReadHolder {
+  public int messageLength;
+  public ByteBuffer messageBuffer;
```
This should be an ArrowBuf, not a ByteBuffer.

BryanCutler replied:
This is for reading the flatbuffer message. It was previous being read into a ByteBuffer and from what I can tell that is the only option for the generated Message constructor.  From flatbuf/Message.java:
```java
public final class Message extends Table {
  public static Message getRootAsMessage(ByteBuffer _bb) { return getRootAsMessage(_bb, new Message()); }
  public static Message getRootAsMessage(ByteBuffer _bb, Message obj) { _bb.order(ByteOrder.LITTLE_ENDIAN); return (obj.__init(_bb.getInt(_bb.position()) + _bb.position(), _bb)); }
  ...
  ```

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReadHolder.java:
```diff
@@ -0,0 +1,34 @@
+ * messageBuffer contains the raw bytes of the message, message is the realized flatbuf Message.
+ */
+public class MessageReadHolder {
+  public int messageLength;
```
The open struct interface is never the best. If we want to use it, we should clarify why in the class.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReader.java:
```diff
@@ -18,48 +18,50 @@
+   * @throws IOException
    */
-  long bytesRead();
+  protected abstract void readMessage(T holder) throws IOException;
```
I don't think we should separate readMessage and readMessageBody as separate methods. The holder can hold both. Let's just have a single boolean readMessage(T holder) that returns false if there are no more messages.

BryanCutler replied:
I'm ok with that, but I believe @icexelloss had an earlier concern that when implementing a custom MessageReader, the subclass would then have to be aware of some details of the Arrow stream format. Specifically, checking for the EOS and if the message is followed by a body or not.

Also, I'm not sure how much value having an interface with just `boolean readMessage(T holder)` since the implementation will have to do everything and we could not use this interface for `ArrowStreamReader` since that requires a `ReadChannel` and returning the message body in an `ArrowBuf`.

BryanCutler replied:
I think what is useful for someone writing a custom `MessageReader` would be to have a hook after reading the message to inspect the message type and then some way to override the reading of the body in case they do not want it to be read into an `ArrowBuf`.  In my case, I need to check for a `RecordBatch` message and then forward the body data somewhere else. I'll think about this a little more..

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReader.java:
```diff
@@ -18,48 +18,50 @@
+ * Abstract base class for reading a sequence of messages.
  */
-public interface MessageReader {
+public abstract class MessageReader<T extends MessageReadHolder> {
```
Let's make the exception generic as well. There are cases where no IOException will be thrown (or the exception will be something else).

BryanCutler replied:
Yeah, good idea

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReader.java:
```diff
@@ -18,48 +18,50 @@
+ * Abstract base class for reading a sequence of messages.
  */
-public interface MessageReader {
+public abstract class MessageReader<T extends MessageReadHolder> {
```
Please make this AutoCloseable with the close throwing the generic parameter suggested above for exception.

BryanCutler replied:
Sure, I can do that

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowBufReadHolder.java:
```diff
@@ -0,0 +1,29 @@
+ * Extend MessageReadHolderBase to contain a message body ArrowBuf used when reading messages
+ * with MessageReader.
+ */
+public class ArrowBufReadHolder extends MessageReadHolder {
```
I don't follow having two different holder implementations. The message should always be in ArrowBufs (see comment on the base MessageReadHolder below).

BryanCutler replied:
This adds an `ArrowBuf` to hold the message body off heap . In my use case, I don't need to load the body to an `ArrowBuf` and doing so would just add extra overhead in copying the data back and forth.  It would be possible to combine these into one holder class, because most people will probably read the message body off heap. In my case, I would just not set this value in the holder.  Does that sound better?  

For example:
```java
public class MessageReadHolder {
  public int messageLength;
  public ByteBuffer messageBuffer;
  public Message message;
  public ArrowBuf bodyBuffer;
}
```

icexelloss commented:
@BryanCutler Please give me a few more days to review since I am still traveling. I will try to take another look on 22nd.

BryanCutler commented:
Thanks for the review and suggestions @jacques-n! I don't think the flatbuffer message can be constructed from an `ArrowBuf`, which is why `ByteBuffer` is used. I'll think about the interface a bit more to see if it makes sense having a single method `readNextMessage`.

BryanCutler commented:
I refactored a little to improve on some things. @jacques-n if you could take another look when you have time, that would be much appreciated!  I still think the `MessageReader` class should have a separate method to read the message body so that a custom reader could handle the body data differently than reading to an `ArrowBuf` if needed. I tried to simplify the classes and I'll sum up the changes below.  Thanks!

BryanCutler commented:
**Update 2**

I removed the MessageHolder classes and replaced with `MessageResult` which is returned when a message is read so that the user does not have to create an instance on each read. If the message has a body, then the class `MessageResultData` will wrap the `MessageResult` and add an object to hold the message body data, using a type param `T`. 

`MessageReaderBase` is an abstract class that includes logic to read messages and populate the `MessageResult` class above. Users would call `public MessageResultData<T> readNext()` to get an object that will indicate if there is a valid message and contains all relevant data.  Implementations only need to define a method to read the message body with the signature `protected abstract T readMessageBody(MessageResult holder)`.

The implementation `MessageReader` defines `protected ArrowBuf readMessageBody(MessageResult result)` which is used by ArrowStreamReader, similar to before.

I think this new design makes it a little more clear what data is being used where and easier to make a custom MessageReader by just defining a type for the message body and a method to read to it.

BryanCutler commented:
@wesm , it looks like this needs some more discussion before it can be merged. I really wanted to get this into 0.10.0 because I plan to update Spark with this version and this would help clean up quite a few details. It's not a hard requirement though, so I don't want to delay the release. Should I remove the target version for this and if it by chance gets resolved in time, then maybe we can fit it in?

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReaderBase.java:
```diff
@@ -0,0 +1,88 @@
+ * Abstract class to define a reader for a sequence of messages using a ReadChannel. Type parameter
+ * T will be used by the implementing class to hold the message body.
+ */
+public abstract class MessageReaderBase<T> implements AutoCloseable {
```
Message Reader should be built for consumers, not MessageReader implementers. Also, it shouldn't require the use of a ReadChannel. Part of the reason that I jumped through hoops in the flight patch is that the reader base assumes you have a readchannel. That's an implementation detail of a particular reader implementation

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReaderBase.java:
```diff
@@ -0,0 +1,88 @@
+   * @return MessageResultData containing the message and message body.
+   * @throws IOException
+   */
+  public MessageResultData<T> readNext() throws IOException {
```
Again, let's not mix a particular implementation with the interface.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageResult.java:
```diff
@@ -0,0 +1,83 @@
+   * @param messageBuffer contains the raw bytes of the message
+   * @param message the realized flatbuf Message
+   */
+  public MessageResult(int messageLength, ByteBuffer messageBuffer, Message message) {
```
Why have message Result and MessageResult data as separate classes? Let's just collapse them into a single class.

The fact that a particular form of a message is backed by a ByteBuffer is probably an implementation detail. Why expose that in MessageResult?

BryanCutler replied:
>Why have message Result and MessageResult data as separate classes?

This was so that I could return all of the information related to reading a flatbuf message, like this:
```java
public static MessageResult readMessage(ReadChannel in) throws IOException {
    ...
      messageLength = MessageSerializer.bytesToInt(buffer.array());

      // Length of 0 indicates end of stream
      if (messageLength != 0) {

        // Read the message into the buffer.
        messageBuffer = ByteBuffer.allocate(messageLength);
        if (in.readFully(messageBuffer) != messageLength) {
        ...
        // Load the message.
        message = Message.getRootAsMessage(messageBuffer);
      }
    }
    return new MessageResult(messageLength, messageBuffer, message);
  }
```
This way, the function produces a very clear output. Otherwise, if returning `MessageResultData<T>` it would need to be generic for no reason.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageResult.java:
```diff
@@ -0,0 +1,83 @@
+  public MessageResult(int messageLength, ByteBuffer messageBuffer, Message message) {
+    this.messageLength = messageLength;
+    this.messageBuffer = messageBuffer;
+    this.message = message;
```
To clarify, I think that a RecordBatch body should be offheap. A message is really just a pojo (that may be backed by a bytebuffer today). On that point, we should probably be create a org.apache.arrow.vector.types.pojo.Message object and use that for our public interfaces, not the flatbuf.

BryanCutler replied:
That seems fine to me to make a `pojo.Message` and not use flatbuf types in our exposed APIs. The only thing is the flatbuf message is already exposed in many of the MessageSerializer functions, so would you mind that we do this as a followup task?

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageResultData.java:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @param <T> Type parameter for an object that holds message body data
+ */
+public class MessageResultData<T> {
```
Why generic? I think a record batch message should always be an ArrowBuf.

BryanCutler replied:
For my use case, I am just inspecting the message type and don't care about processing any data. I also can not store the data off heap, so any usage of an ArrowBuf would then need to immediately copy it again, which isn't ideal. This is part of a constraint with working within Spark and probably not a common case. I agree that most cases would probably use an ArrowBuf, so would it be preferable to make this non-generic and include an ArrowBuf? As long as there is a way to override it for my needs, I'm fine with that.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -381,30 +473,29 @@ public static ArrowDictionaryBatch deserializeDictionaryBatch(ReadChannel in,
-  public static ArrowMessage deserializeMessageBatch(MessageReader reader, BufferAllocator alloc) throws IOException {
-    Message message = reader.readNextMessage();
-    if (message == null) {
+  public static ArrowMessage deserializeMessageBatch(MessageReader reader) throws IOException {
```
I think this is the only public method here that takes `MessageReader` as args? Seems a bit weird, any particular reason for that?

BryanCutler replied:
Yeah, this doesn't really fit with the rest of the APIs and I didn't want to change it too much because I wasn't sure if someone relied on it.  @elahrvivaz , do you by chance use this one?

The problem with using `MessageReader` as an arg in the other functions is that they deserialize only a specific message type, e.g. `deserializeDictionaryBatch`, and `MessageReader` is not aware of any order of the messages - that is the job of `ArrowStreamReader`.  So to me it makes more sense to have either a `ReadChannel` or raw message data as input to deserialize.

elahrvivaz replied:
We only use the top-level reader API, so the internals don't matter much to me. Basically just `reader.loadNextBatch` and then use the populated `VectorSchemaRoot`

BryanCutler commented:
Thanks @jacques-n , iiuc I think you are looking for a pure interface definition, which is fine with me. I would still like something that I could extend which encapsulates the message stream handling details, but let me try to define an interface without mixing any particular implementation.

BryanCutler commented:
Here is a proposal with an interface that is free from any implementations (let me know it's preferable to have all this in a google doc to discuss).  Apologies for backtracking a little, but I put back in the MessageHolder class because my implementation ended up quite awkward otherwise. @jacques-n let me know if you still prefer not to use a holder class this way.

**Interface**

```java
class MessageHolder {
  public Message message;
}

interface MessageReader<T extends MessageHolder, U extends Exception> extends AutoCloseable {

  boolean readNext(T holder) throws U;

  void close() throws U;
}
```

**Partial Implementation using ReadChannel**
This does define how to read a message body, but encapsulates stopping at the EOS and reading the body data only in the body length is > 0.

```java

// Adds the raw message buffer to MessageHolder
class MessageChannelHolderBase extends MessageHolder {
  public int messageLength;
  public ByteBuffer messageBuffer;
}

abstract class MessageChannelReaderBase<T extends MessageChannelHolderBase> implements MessageReaderInterface<T, IOException> {
  protected ReadChannel in;

  public MessageChannelReaderBase(ReadChannel in) {
    this.in = in;
  }

  @Override
  public boolean readNext(T holder) throws IOException {
    MessageResult result = MessageSerializer.readMessage(in);

    if (!result.hasMessage()) {
      return false;
    }

    holder.message = result.getMessage();
    holder.messageBuffer = result.getMessageBuffer();
    holder.messageLength = result.getMessageLength();
    handleMessage(holder);

    return true;
  }

  @Override
  public void close() throws IOException {
    in.close();
  }

  // Hook to override handling after reading a message, default behavior is to read body
  protected void handleMessage(T holder) throws IOException {
    if (holder.message.bodyLength() > 0) {
      readMessageBody(holder);
    }
  }

  protected abstract void readMessageBody(T holder) throws IOException;
}
```

**Implementation with ReadChannel**
This reads the message body from a ReadChannel into an ArrowBuf

```java
class MessageChannelHolder extends MessageChannelHolderBase {
  ArrowBuf bodyBuffer;
}

class MessageChannelReader extends MessageChannelReaderBase<MessageChannelHolder> {
  protected BufferAllocator allocator;

  public MessageChannelReader(ReadChannel in, BufferAllocator allocator) {
    super(in);
    this.allocator = allocator;
  }

  @Override
  protected void readMessageBody(MessageChannelHolder holder) throws IOException {
    holder.bodyBuffer = MessageSerializer.readMessageBody(in, holder.message.bodyLength(), allocator);
  }
}
```

**Alternatives**
* Add ArrowBuf into `MessageHolder` since this is the most common case, and my implementation would just not use it.  I could eliminate the `MessageChannelReaderBase` then, but keeping it would be my preference.

* It might not be necessary to keep `messageLength` and `messageBuffer` in the `MessageChannelHolder` because I think these can be accessed from the flatbuf.Message. But since they are being created in MessageSerializer.readMessage anyway, I don't think it hurts to keep them so that would be my preference.

Thanks in advance for reviewing!

wesm commented:
Thanks @BryanCutler. Since we're so close to 0.10.0 I'm going to check in roughly daily to make sure this keeps making progress. If it seems like the work is being rushed then we should release 0.10.0 without this

jacques-n commented:
@BryanCutler , can you expound on your usecase where you aren't using ArrowBuf?

wesm commented:
To be clear, my understanding of the above is that the `ByteBuffer` used above would only hold a Flatbuffer, not the message body. @BryanCutler is that not the case?

jacques-n commented:
I think there are two things: messages and data bodies. I consider them two parts of the a single object. I'm proposing that the interface get those back together. The message is generally small and should be treated as a pojo. The data body is larger and should be using ArrowBufs. Thus my question of in what situation ArrowBufs would be insufficient/wrong for the data body. (Let me know if I'm missing something here.)

jacques-n commented:
Per my comment above, if the message happens to use a ByteBuffer, that should probably be the concern of the message object. And we can make message objects autocloseable if there is a situation where the message objects can't rely on gc.

BryanCutler commented:
>To be clear, my understanding of the above is that the ByteBuffer used above would only hold a Flatbuffer, not the message body

Yes, `MessageChannelHolder` base contains the result of reading a flatbuffer Message object, not the message body.  This include the length of the message read, the ByteBuffer the message was read into, and the realized `Message` object.

BryanCutler commented:
> @BryanCutler , can you expound on your usecase where you aren't using ArrowBuf?

My specific use case is that I am trying to read an Arrow stream in Spark and parallelize the record batches so that they are distributed in the Spark cluster. I have to work within the existing mechanism to parallelize, so that puts the constraint that the record batch needs to be a byte array. These byte arrays exist for the duration of the DataFrame and trying to change this to use an `ArrowBuf` and `BufferAllocator` would require significant changes. It might be possible to change this in the future, because it probably would be better to use off heap memory, but for now I need to make it work under that constraint.

So to process the stream, I need to read the flatbuf Message and inspect the Message type.  If it is a RecordBatch, then I can look at the body length and create a byte array that will fit the message+body.  If I have the raw buffer of the flatbuf Message, I can copy that over to the byte array (should be a small copy). Finally, I then read the message body into the byte array and have a complete RecordBatch message.  I need to avoid making copies of the the message body data, because that will be expensive.

Hopefully that makes sense. I think it's probably not common to read the message body this way, but I think it's likely users would read a Message (without body) to inspect it and do some action before reading the message body.

BryanCutler commented:
**Alternative, bare-min design**

This simplifies the classes and holders so that an ArrowBuf is always included. Not my preference, but at least I could override the default usage and just not use the ArrowBuf.  This removes the `handleMessage` hook, which isn't necessary (maybe useful?) and also assumes I can get the correct message length and ByteBuffer from the flatbuf.Message itself, which I'm not 100% sure about.

 **Interface**

```java
class MessageHolder {
  public Message message;
  public ArrowBuf bodyBuffer;
}

interface MessageReader<T extends MessageHolder, U extends Exception> extends AutoCloseable {

  boolean readNext(T holder) throws U;

  void close() throws U;
}
```

**Implementation with ReadChannel**

```java
class MessageChannelReader<T extends MessageHolder> implements MessageReader<T, IOException> {
  protected ReadChannel in;
  protected BufferAllocator allocator;

  public MessageChannelReaderBase(ReadChannel in, BufferAllocator allocator) {
    this.in = in;
    this.allocator = allocator;
  }

  @Override
  public boolean readNext(T holder) throws IOException {
    holder.message = MessageSerializer.readMessage(in);

    if (message == null) {
      return false;
    }

    if (holder.message.bodyLength() > 0) {
      readMessageBody(holder);
    }

    return true;
  }

  @Override
  public void close() throws IOException {
    in.close();
  }

  protected void readMessageBody(T holder) throws IOException {
    holder.bodyBuffer = MessageSerializer.readMessageBody(in, holder.message.bodyLength(), allocator);
  }
}
```

jacques-n commented:
@BryanCutler, I'm okay with the new interface and the new the generics for base MessageHolder. Just wanted to make sure it is was a real use case. In general, I don't think we should try to support on-heap arrow batches as it would be a massive amount of effort.

jacques-n commented:
Why would MessageChannelReaderBase not be ArrowBuf based? ReadChannel basically is.

BryanCutler commented:
Thanks @jacques-n , I agree that we should encourage or support on-heap batches - this is probably a pretty isolated case.

>Why would MessageChannelReaderBase not be ArrowBuf based? ReadChannel basically is.

Yes, but `ReadChannel` also reads into a ByteBuffer, which is what I can use.  Ideally, I would like to use a `ReadableByteChannel` in `MessageChannelReaderBase`, but this makes a problem since `ReadChannel` wraps a `ReadableByteChannel` instead of extending it.  So to be able to do this with `MessageChannelReader` I would need to add a method `ReadChannel.getReadableByteChannel()` to return the wrapped channel.  Does that sound ok to you?

jacques-n commented:
I think we're trying to support two many different situations with these interfaces. Ideally, ReadChannel should be made more generic. Exposing the internal objects makes it less generic. Can you point at the consumer of this code you're trying to customize for? We should decide if we just have two different interfaces.

BryanCutler commented:
I think you're right @jacques-n , this should be kept simple and it doesn't make sense to try to support too so many things. Here is what I tried that seems to be a little better:

* I got rid of `MessageChannelReaderBase` because you're right that `ReadChannel` is pretty much based on ArrowBuf and it's overkill to make an additional interface for a `ReadableByteChannel`.

* I simplified `MessageChannelReader` to just have a `readNext` method, like you suggested before.

* I kept the class `MessageChannelResult` that is returned when reading a flatbuf Message from a ReadChannel. This wraps some of the specifics in dealing with the flatbuf Message, and I think we can probably replace with with a pojo Message, which I will create a followup JIRA to do.

With the above changes, I can implement my own MessageReader, and make use of the `MessageChannelResult` and other static function in MessageSerializer to make it work without dealing with the low level details.  Hopefully, this sounds ok still.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -440,24 +531,58 @@ public static ByteBuffer serializeMessage(FlatBufferBuilder builder, byte header
+   * @return MessageResult with Message and message information
+   * @throws IOException
+   */
+  public static MessageChannelResult readMessage(ReadChannel in) throws IOException {
```
feels like this should be on MessageChannelResult, not MessageSerializer.

BryanCutler commented:
Hey @jacques-n , that's fine. I don't think we should define an interface if we are not sure it will be useful.  I only added that because I thought that's what you were requesting. The only things helpful to my use case are really just the static functions added here that hide some of the low level stream handling. I'll cut this down to only the things that are useful to me or I felt made the API's more consistent.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageHolder.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Class to hold a Message and body when reading messages through a MessageChannelReader.
+ */
+public class MessageHolder {
```
I am a bit confused about `MessageHolder` and `MessageChannelResult`, can you explain a bit what both are for and why do we need both?

BryanCutler replied:
If we have a `MessageChannelReader`, I think that there should a single public method that will read a message + body. `MessageHolder` contains the message and body, whether this is passed by reference or returned by `readNext` doesn't matter to me.

`MessageChannelResult` is the returned object for the low-level function `readMessage` that just reads a flatbuffer message. This class has the message plus the raw data buffer used during reading. The raw data is useful for my case, but probably not most cases, which is why I kept it only in this low-level function.

icexelloss replied:
I see. `MessageHolder` seems fine to me because it's just contains two parts (header + body) for a message. `MessageChannelResult` does feel a bit arbitrary to me. Does `Message.getByteBuffer()` doesn't do what you want?

BryanCutler replied:
`Message.getByteBuffer()` returns the same buffer right now because of the way it's created, but then I would be reliant on that assumption. If something changed internally in this function, then this might not be true. So by returning the raw byte buffer of the message makes a guarantee of the contents.

icexelloss replied:
Sorry for the blocking the PR at such time...But want to make sure we don't introduce something unnecessary here...

BryanCutler replied:
> Do you mean that it's possible that Message.getByteBuffer() has different content than in MessageChannelResult.messageBuffer

I'm saying from what I can tell, there is no guarantee that `Message.getByteBuffer()` only contains the message bytes, and it's possible there could be additional bytes before or after. When the `Message` is created, they do some accounting with the buffer position and that is not public.

icexelloss replied:
Fair enough. I did some googling but didn't find a good answer either.

wesm replied:
@BryanCutler @icexelloss see my comments above; the lack of clarity in the comments to help distinguish the serialized metadata from the message body probably fueled this long comment thread

icexelloss commented:
This PR looks generally OK to me. I am not a big fan of the `MessageChannelResult` (seems not ideal to have a class just to hold the deserialized bytes of a flatbuf message) and have methods return just the flatbuf message is a cleaner API. But I am not going to block this PR because of this either. Hopefully we have chance to clean this in the future if there turns out to be a better way.

BryanCutler commented:
Thanks @icexelloss , I agree that it could probably be improved and maybe use a pojo Message instead of exposing flatbuf in our public APIs. So do you approve this PR to be merged?

wesm commented:
I'm going to have a look through this so I can understand a bit better some of the controversy. We have similar classes for dealing with message streams in C++

wesm commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelResult.java:
```diff
@@ -0,0 +1,104 @@
+  * Construct a container to hold a message result.
+  *
+  * @param messageLength the length of the message read in bytes
+  * @param messageBuffer contains the raw bytes of the message
```
I think you should be clear here that the "message" is the message _metadata_, and does not include the body (this can be inferred by reading onward of course...). Flatbuffers uses `ByteBuffer` as it binary interface is my understanding (I have confirmed this by looking at the .java files generated by flatc), and so this will help prevent possible confusion why there is ByteBuffer here instead of ArrowBuf

wesm commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelResult.java:
```diff
@@ -0,0 +1,104 @@
+import org.apache.arrow.flatbuf.Message;
+
+/**
+* Class to hold resulting Message and message information when reading messages from a ReadChannel.
```
Mention here that this class does not interact with the message body, only the serialized Flatbuffer message description

wesm commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageSerializer.java:
```diff
@@ -320,33 +382,50 @@ public static ArrowBlock serialize(WriteChannel out, ArrowDictionaryBatch batch)
-  public static ArrowDictionaryBatch deserializeDictionaryBatch(MessageReader reader,
-                                                                Message message,
-                                                                BufferAllocator alloc) throws IOException {
+  public static ArrowDictionaryBatch deserializeDictionaryBatch(Message message, ArrowBuf bodyBuffer) throws IOException {
```
At 122 characters long this is surely violating the Java style guide. The sooner we can fail the build because of style issues, the better. We need to keep the codebase clean @siddharthteotia @jacques-n @icexelloss @BryanCutler

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowStreamReader.java:
```diff
@@ -101,19 +98,23 @@ protected void closeReadSource() throws IOException {
 
     // Reached EOS
-    if (message == null) {
+    if (!messageReader.readNext(holder)) {
```
I agree with having immutablility. This was partly residual from some of the previous rounds, but there is no real reason to keep this immutable here.

BryanCutler commented:
Thanks @wesm and all others who reviewed! I'll create a followup to address your comments.
